src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+                                        ::dynamic/return
+                                        ::dynamic/request-decision
+                                        ::dynamic/request-comment}})))
+
+(defmethod application-view-specific :application.event/returned
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/comment-requested
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/commented
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/decision-requested
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/decided
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/approved
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/rejected
+  [application event]
+  application)
+
+(defmethod application-view-specific :application.event/closed
+  [application event]
+  application)
+
+(deftest test-application-view-specific
+  (testing "supports all event types"
+    (is (= (set (keys dynamic/event-schemas))
+           (set (keys (methods application-view-specific)))))))
+
+
+(defn- log-event [events event]
+  ;; only include events which are about processing the application
+  (if (contains? #{:application.event/created
+                   :application.event/draft-saved}
+                 (:event/type event))
+    events
+    ;; choose only keys which will be shown in the UI's event log
+    (conj events (select-keys event [:event/type
+                                     :event/time
+                                     :event/actor
+                                     :application/comment
+                                     :application/commenters
+                                     :application/decider
+                                     :application/decision]))))
+
+(defn- application-view-generic
+  "See `application-view`"
+  [application event]
+  (assert (= (:application/id application)
+             (:application/id event))
+          (str "event for wrong application "
+               "(not= " (:application/id application) " " (:application/id event) ")"))
+  (-> application
+      (assoc :application/last-activity (:event/time event))
+      (update :application/events log-event event)))
+
+;; TODO: replace rems.workflow.dynamic/apply-event with this
+;;       (it will couple the write and read models, but it's probably okay
+;;        because they both are about a single application and are logically coupled)
+(defn- application-view
+  "Projection for the current state of a single application.
+  Pure function; must use `assoc-externals` to enrich the model with
+  data from other entities."
+  [application event]
+  (-> application
+      (application-view-specific event)
+      (application-view-generic event)))
```

Comment:
I think these functions could really be named better. How about `event-application-view` for the multimethod and just more specific fns for the generic parts?

Reply:
Splitting the one generic handler for all events to multiple handlers did not improve things:

```clj

(defn- application-id-check-application-view
  [application event]
  (assert (= (:application/id application)
             (:application/id event))
          (str "event for wrong application "
               "(not= " (:application/id application) " " (:application/id event) ")"))
  application)

(defn- event-log-application-view
  [application event]
  (update application :application/events append-to-event-log event))

(defn- last-activity-application-view
  [application event]
  (assoc application :application/last-activity (:event/time event)))

(defn- application-view
  "Projection for the current state of a single application.
  Pure function; must use `assoc-injections` to enrich the model with
  data from other entities."
  [application event]
  (-> application
      (event-application-view event)
      (application-id-check-application-view event)
      (event-log-application-view event)
      (last-activity-application-view event)))
```

Reply:
The name `event-application-view` is also confusing, because *all* the functions are applying events to the application model. The only difference that the multimethod is specific to some event types, whereas the other function contains generic functionality which applies to all events.

Reply:
Another option:

```clj
(defn- application-view
  "Projection for the current state of a single application.
  Pure function; must use `assoc-injections` to enrich the model with
  data from other entities."
  [application event]
  (assert (or (nil? (:application/id application))
              (= (:application/id application)
                 (:application/id event)))
          (str "event for wrong application "
               "(not= " (:application/id application) " " (:application/id event) ")"))
  (-> application
      (event-specific-application-view event)
      (assoc :application/last-activity (:event/time event))
      (update :application/events append-to-event-log event)))
```

Reply:
I like this `event-specific-application-view`. Could also be `event-type-specific-application-view`.

And the other "generic" code looks fine now, it's not hidden under the functions. When it gets more complicated, we can add named functions to do the specific common parts.

Reply:
Sounds good. I'll do it like in the last version.